Add support for `ready_for_review` GitHub PR event

nickd-kt opened the pull request:
# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

Add support for a GitHub PR event `ready_for_review`, which is triggered when a PR is converted from draft to active.

Resolves: https://github.com/openshift-pipelines/pipelines-as-code/issues/1953

# Submitter Checklist

- [ ] 📝 Ensure your commit message is clear and informative. Refer to the How to write a git commit message guide. Include the commit message in the PR body rather than linking to an external site (e.g., Jira ticket).

- [ ] ♽ Run make test lint before submitting a PR to avoid unnecessary CI processing. Consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the repository root for an efficient workflow.

- [ ] ✨ We use linters to maintain clean and consistent code. Run make lint before submitting a PR. Some linters offer a --fix mode, executable with make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) are installed).

- [ ] 📖 Document any user-facing features or changes in behavior.

- [ ] 🧪 While 100% coverage isn't required, we encourage unit tests for code changes where possible.

- [ ] 🎁 If feasible, add an end-to-end test. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for details.

- [ ] 🔎 Address any CI test flakiness before merging, or provide a valid reason to bypass it (e.g., token rate limitations).

- If adding a provider feature, fill in the following details:

| Git Provider     | Supported |
|------------------|-----------|
| GitHub App       | ❌️        |
| GitHub Webhook   | ✅️        |
| Gitea            | ❌️        |
| GitLab           | ❌️        |
| Bitbucket Cloud  | ❌️        |
| Bitbucket Server | ❌️        |

nickd-kt committed:
Add support for ready_for_review GitHub PR event

chmouel commented:
thanks for your contribution

it's look good to me, i just need to test it manually and see if there is any issues and perhaps add a e2e (not asking you to do it, because it's not trivial to get it started)

chmouel commented:
/ok-to-test

chmouel committed:
docs: clarify github trigger behavior

Update the running.md guide to better explain when PipelineRuns are
triggered, specifically mentioning that draft PRs require "ready for
review" status before triggering. Also reorganized content by moving
GitHub App info from the hint box to the main text for better
visibility.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
I have tested it and it works nicely! I have pushed to your PR to update the documentation a bit, let me know if that looks good to you and i think we are good to go 🚀

nickd-kt commented on docs/content/docs/guide/running.md:
```diff
@@ -24,8 +24,20 @@ For example, if a PipelineRun has this annotation:
+permissions (see below).
+
+When using GitHub as a provider, Pipelines-as-Code will not run on draft Pull Requests until
+you set the Pull Request as ready for review.
```
From my experience, PAC **does** run on draft PRs when you push a new commit or just update the PR to the latest `main` / `master` branch (both produce the `pull_request` event with `synchronize` action)

However, users can opt out from running pipelines on draft PRs by adding `!body.pull_request.draft` in the `on-cel-expression` annotation:
```
pipelinesascode.tekton.dev/on-cel-expression: event == "pull_request" && !body.pull_request.draft
```

nickd-kt replied:
Given the above, I'd change the language to something like (sorry you'll need to remove the slashes `\` in the code block):

```suggestion
When using GitHub as a provider, Pipelines-as-Code runs on draft Pull Requests by default. However, you can prevent pipelines from triggering on draft Pull Requests by using the following annotation:

\```yaml
pipelinesascode.tekton.dev/on-cel-expression: event == "pull_request" && !body.pull_request.draft
\```

With this configuration, your pipeline will only be triggered when the Pull Request is converted to "Ready for Review." For additional examples, see [Advanced event matching using CEL](https://pipelinesascode.com/docs/guide/matchingevents/#advanced-event-matching-using-cel).
```

chmouel replied:
oh really, thought github didn't send events.... but i guess if we didn't run on draft i would have heard about it from users :) 

thanks for the clarification and the example

nickd-kt replied:
absolutely!

nickd-kt commented on docs/content/docs/guide/running.md:
```diff
@@ -24,8 +24,20 @@ For example, if a PipelineRun has this annotation:
-long as the submitter is allowed to run it.
+When a Pull Request is created, the PipelineRun will be automatically triggered
+and executed, provided that the person who submitted it has the appropriate
+permissions (see below).
```
Looks like the sentence starts just above the block code, so this should be 

```suggestion
it will be automatically triggered and executed when a user with appropriate permissions submits a Pull Request. See ACL Permissions for running on a event below.
```

chmouel replied:
I am bit lost with the edit 😅  feel free to just push --force-with-lease over my commit to your branch :) so i can approve

nickd-kt committed:
Update docs/content/docs/guide/running.md

nickd-kt committed:
Update docs/content/docs/guide/running.md

nickd-kt committed:
Remove slashes from markdown code block

nickd-kt committed:
Merge branch 'main' into nick-d/ready-for-review

nickd-kt committed:
Fix grammar in the doc

nickd-kt commented:
@chmouel now it's ready for review. Appreciate your help!

chmouel commented:
perfect thanks

chmouel commented:
/test linters

chmouel commented:
(will squash the commit before merge)

chmouel commented:
/lgtm

chmouel commented:
/merge

chmouel commented:
/merge
